### PR TITLE
Store server timezone before setting to UTC

### LINKED ISSUE #8212: Feature Request: add 'logtimezone' automatically to config.php during install
It would be sensful if during the installation / finalisation of owncloud, the timezone of the server is added to config php automatically.
Logs created by oC would then be with the right timzone from beginning.

### PR DESCRIPTION
**Pull request by RobinMcCorkell**
Used for the default log timezone. Fixes #8212

See #14898 for the previous incarnation. This version doesn't remove the default timezone to UTC setting, so it won't break anything.

cc @karlitschek @DeepDiver1975 @PVince81

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
But what is the benefit now of this change? Sorry - I don't get it :speak_no_evil:

**Comment by Xenopathic** [comment]
@DeepDiver1975 It allows oC to get the timezone configured on the server, which is a good default for many things (default log timezone is the use case here). It could conceivably be used to get the default user timezone too, but it's not hooked up yet.

**Comment by karlitschek** [comment]
:+1:

**Comment by Xenopathic** [comment]
Oh, this would also allow #13733 to function as it should (with the code being changed to use `getServerDateTimeZone()`, of course). In fact, #13733 would effectively become redundant, since this makes the default timezone if no logtimezone is set to be the server timezone, which is the effective functionality of that change.

**Comment by Xenopathic** [comment]
I removed the setting of `logtimezone` during the setup, as the default timezone for logs is the server timezone with this PR anyway, so it is redundant. cc @mmattel

**Comment by MorrisJobke** [comment]
Isn't this fixed with: #13733 cc @DeepDiver1975

**Comment by Xenopathic** [comment]
@MorrisJobke No, #13733 never actually worked properly, since the PHP default timezone gets set to UTC in base.php before that code (which gets the default timezone) runs. Thus the only `logtimezone` that will ever be written to config.php is UTC. This PR allows any code in oC to retrieve the system timezone, and the usecase here is for the default timezone when a `logtimezone` isn't set. As #13733 would have conflicted with this (UTC being set during setup) I removed the changes from that.

**Comment by MorrisJobke** [comment]
I don't really like to play with the timezone... and I don't see a valid use case for this. It's totally fine to have the log in UTC. Just my 2 ct.

**Comment by mmattel** [comment]
Then the config.php parameter ``logtimezone`` would not necessary anymore, because when owncloud sets this internally to UTC, everything gets logged in UTC.

But I guess, that there are people around who want to have the log in their natural timezone and link it with other logs like from the webserver.

My attempt was, to writeout the timezone used to config.php, so it is for admins easier changing it.
Xenopathic approach is fine as he directly uses the server timzone set including a fallback.

The only suggestion here would be, to additionally write once the timzone gathered by oC to config.php

**Comment by MorrisJobke** [comment]
@Xenopathic rebase needed

**Comment by PVince81** [comment]
Can we make this happen at setup time ? In setup.php, detect the timezone then write it into config.php.

I see the code changes base.php and other locations so I have the feeling it would run more often than necessary. But I didn't dig deeper so might have misunderstood.

**Comment by Xenopathic** [comment]
@PVince81 We don't have much choice about touching base.php - in the init routine, the timezone is set to UTC, and we need to extract that before base.php is included at all (which is too early to do any meaningful work)

File: lib/base.php [hunk 1]
```diff
@@ -491,6 +491,11 @@ public static function init() {
 			ini_set('display_errors', 1);
 		}
 
+		try {
+			$serverDateTimeZone = new \DateTimeZone(date_default_timezone_get());
+			\OC::$server->registerParameter('ServerDateTimeZone', $serverDateTimeZone);
+		} catch (\Exception $e) {
+		}
```
Comment (PVince81) [inline lib/base.php]: Add comment why we are not handling exception ?

File: lib/private/log/owncloud.php [hunk 2]
```diff
@@ -61,11 +61,16 @@ public static function write($app, $message, $level) {
 		if($level>=$minLevel) {
 			// default to ISO8601
 			$format = OC_Config::getValue('logdateformat', 'c');
-			$logtimezone=OC_Config::getValue( "logtimezone", 'UTC' );
+			$timezone = null;
 			try {
-				$timezone = new DateTimeZone($logtimezone);
+				$logtimezone = OC_Config::getValue("logtimezone");
+				if ($logtimezone) {
```
Comment (PVince81) [inline lib/private/log/owncloud.php]: `$logtimezone !== null` ? (such strings checks can be dangerous as PHP likes to evaluate strings like "0" to true ...)

**Comment by PVince81** [comment]
I like the general idea but I wonder how it works in general, whether admins prefer UTC by default or the local timezone. I see systemd shows me the log in the current timezone.

**Comment by Xenopathic** [comment]
@PVince81 @DeepDiver1975 What's your consensus on this? Storing logs in UTC by default isn't a useful default, and we should use the system timezone IMHO.

**Comment by PVince81** [comment]
The logs are stored in JSON which isn't very readable in some cases either... it seems the logs were originally designed to be read from a log viewer. Assuming that it should be fine to have the log times as UTC as they can be converted to viewer time with a log viewer.

Cool would be to have a command "./occ log" that parses and shows the log in a friendly local format :smile:

**Comment by Xenopathic** [comment]
@PVince81 Actually, that's a better idea. So log timezone should always be UTC (unless the admin specifically changes it).

As for `./occ log`... I was thinking of implementing a couple of log management commands. Once files_external is sane again, I might get on that.

**Pull request closed without merging**